### ISSUE TITLE
All yamllint violations have the same generic description in codeclimate reports

### ISSUE THREAD

**Issue by duncanmmacleod** [ansible/ansible-lint#2106]
##### Summary

When running `ansible-lint -f codeclimate` all violations reported up from `yamllint` are printed with the same generic description:

> Rule violations reported by YamlLint when this is installed.\n\nYamllint configuration will be loaded following yamllint's configuration file\nresolution order.\n\nYou can fully disable all of them by adding 'yaml' to the 'skip_list'.\n\nSpecific tag identifiers that are printed at the end of rule name,\nlike 'trailing-spaces' or 'indentation' can also be be skipped, allowing\nyou to have a more fine control.

This means that when reported on gitlab merge request UIs (for example) all violations look the same and there isn't a view towards the actual violation (one could argue that's a failure of the gitlab interface).

It would be better (IMO) to copy the `problem.desc` received from `yamllint` into the `description` field, if that's possible and convenient.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
$ ansible --version
ansible [core 2.12.5]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/duncan/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/duncan/opt/mambaforge/envs/py310/lib/python3.10/site-packages/ansible
  ansible collection location = /home/duncan/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/duncan/opt/mambaforge/envs/py310/bin/ansible
  python version = 3.10.4 | packaged by conda-forge | (main, Mar 24 2022, 17:38:57) [GCC 10.3.0]
  jinja version = 3.0.3
  libyaml = True
$ ansible-lint --version
ansible-lint 6.0.2 using ansible 2.12.5
```

- ansible installation method: conda-forge
- ansible-lint installation method: conda-forge

##### OS / ENVIRONMENT

```console
$ conda list ansible
# packages in environment at /home/duncan/opt/mambaforge/envs/py310:
#
# Name                    Version                   Build  Channel
ansible-compat            2.0.2              pyhd8ed1ab_0    conda-forge
ansible-core              2.12.5             pyhd8ed1ab_0    conda-forge
ansible-lint              6.0.2              pyhd8ed1ab_0    conda-forge
```
##### STEPS TO REPRODUCE

```bash
mkdir -p tmp/tasks
cat > tmp/tasks/main.yml <<EOF
- name: remove file
  ansible.builtins.file:
    path: /tmp/something.txt
    state:  absent
EOF
ansible-lint tmp/ -f codeclimate | python -m json.tool
```

##### Desired Behavior

Report an informative `description` in codeclimate output

##### Actual Behavior

Please give some details of what is actually happening.
Include a [minimum complete verifiable example] with:

- playbook
- output of running ansible-lint
- if you're getting a stack trace, output of
  `ansible-playbook --syntax-check playbook`

```paste below
$ ansible-lint tmp/ -f codeclimate | python -m json.tool
You can skip specific rules or tags by adding them to your configuration file:
# .config/ansible-lint.yml
warn_list:  # or 'skip_list' to silence them completely
  - yaml  # Violations reported by yamllint.

Finished with 1 failure(s), 0 warning(s) on 2206 files.
[
    {
        "type": "issue",
        "check_name": "[yaml] too many spaces after colon",
        "categories": [
            "formatting",
            "yaml"
        ],
        "severity": "major",
        "description": "Rule violations reported by YamlLint when this is installed.\n\nYamllint configuration will be loaded following yamllint's configuration file\nresolution order.\n\nYou can fully disable all of them by adding 'yaml' to the 'skip_list'.\n\nSpecific tag identifiers that are printed at the end of rule name,\nlike 'trailing-spaces' or 'indentation' can also be be skipped, allowing\nyou to have a more fine control.\n",
        "fingerprint": "5a0beb4009b77fa013a3d0ee3b3825084d367f010fe0e77df86bfed3993cdcaf",
        "location": {
            "path": "tmp/tasks/main.yml",
            "lines": {
                "begin": 4
            }
        }
    }
]
```

[minimum complete verifiable example]: http://stackoverflow.com/help/mcve

**Comment by ssbarnea** [comment]
A PR would be welcomed here. Does this also mean that gitlab fails to display the check_name in any way?

Description is supposed to add extra information to the brief error message (which goes into check_name).

**Comment by duncanmmacleod** [comment]
> Does this also mean that gitlab fails to display the check_name in any way?

This is what GitLab displays in its [code quality widget](https://docs.gitlab.com/ee/user/project/merge_requests/code_quality.html#code-quality-widget) for a merge request:

![image](https://user-images.githubusercontent.com/1618530/168305409-e6765d3b-61eb-4f31-a1ca-b3e6c46059f1.png)

which is ... unhelpful.

**Comment by ssbarnea** [comment]
The more I read about this bug the more I think that is a problem with GitLab implementation as they failed to display the `shortdesc/check_name` and only printed `description` field.

If we tune the two json outputs (`sarif` and `codeclimate`) to use the same text for shortdesc and description, we will likely cripple other consumers that might be able to properly make use of both.

I want to fix it but I am really not sure which direction we should take. Probably a link to specifications might clarify how these fields are supposed to be used or specific examples how other consumers are treating these.

In the end, the current purpose of `description` field in rules is to document them, and usually that means that it contains markdown documentation and a level of verbosity that is likely undesired for inclusion.

**Comment by duncanmmacleod** [comment]
> The more I read about this bug the more I think that is a problem with GitLab implementation as they failed to display the shortdesc/check_name and only printed description field.

I agree. I will see if I can report this upstream to gitlab (as well as here).

> In the end, the current purpose of description field in rules is to document them, and usually that means that it contains markdown documentation and a level of verbosity that is likely undesired for inclusion.

I'm not sure I agree 100% with this, at least not in the specific context of codeclimate. The code climate [spec](https://github.com/codeclimate/platform/blob/master/spec/analyzers/SPEC.md#data-types) describes the `description` field as:

    A string explaining the issue that was detected.

There is a second attribute `content` that seems to fit the generic yamllint text better

    A markdown snippet describing the issue, including deeper explanations and links to other resources.

So, perhaps leaning on gitlab to show `check_name` (as well as `description`) **and** improving the `description` field in the codeclimate output here, would be the best of both worlds.

**Comment by ssbarnea** [comment]
Fixing this correctly will need lots of changes but I already started working on them. Basically we will have to convert documentation to markdown and ensure that rules have 3 fields: `shortdesc`, `description` and `help`.

Fist two might vary per instance, but `help` would be constant/generic per rule class. That is important for meta rules, like yaml and schema, which can produce various kind of violations.

At this moment our `description` does not vary, thus this will become the `help` one. Probably we will only copy the shortdesc to description for most rules.

I am not yet sure which terminology we should pick for these 3 fields but I have these options in mind so far:

1. `shortdesc`, `description`, `help`
2. `shortdesc`, `detail`, `description`

# SARIF 

Based on https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10541076 (3.27) I am inclined to say that:

- `result.ruleId` should be our `rule.id` like `schema`
- `result.message` should be `rule.shortdec`
- `result.rule` (see below)

For the verbose `help` part, the one likely to contain reasoning, examples and possible fixed and workarounds we should probably use [reportingDescriptorReference](https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10541300) -- if I read the spec correctly, these are to be listed separately

I will likely need few others to look over the spec as it is far from a clear one.

# CODECLIMATE

Based on 
https://github.com/codeclimate/platform/blob/master/spec/analyzers/SPEC.md that is far more simple than SARIF with `check_name`, `description` and `content` (optional). What they fail to give is a `fingerprint` example. Is not clear to me if that fingerprint is supposed to be a rule-id or something unique like a tuple of (file:line:rule-id)?

Interesting note is that they mention that description `should be fewer than 70 characters long`, which rules out our current help-like description, where most of the time we expect to be at least one paragraph.

`check_name` is defined as "A unique name representing the static analysis check that emitted this issue." - does this mean a sentence or more like rule-id?

**Comment by cidrblock** [comment]
I find myself wondering if long term there is a reason to support a user provided json template as a jinja template.  If there was ever a need for a 3rd variant, or someone wanted something really custom, restructuring the output with a j2 template would be pretty straightforward as long as we had the internally available field names available and well documented

so it would end up looking like:

ansible-lint --output=json --variant=codeclimate
ansible-lint --output=json --variant=sarif
ansible-lint --output=json --varient=mycustom.json.j2

This is _ansible_-lint after all.........  templating with jinja shouldn't be unexpected :)
